microprofile/server/src/main/java/io/helidon/microprofile/server/ServerCdiExtension.java

```diff
@@ -16,56 +16,430 @@
 import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.BeforeDestroyed;
+import javax.enterprise.context.Initialized;
+import javax.enterprise.context.spi.CreationalContext;
 import javax.enterprise.event.Observes;
-import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.DeploymentException;
 import javax.enterprise.inject.spi.Extension;
-import javax.enterprise.inject.spi.ProcessInjectionTarget;
-import javax.ws.rs.Path;
-import javax.ws.rs.core.Application;
+
+import io.helidon.common.HelidonFeatures;
+import io.helidon.common.HelidonFlavor;
+import io.helidon.common.Prioritized;
+import io.helidon.common.configurable.ServerThreadPoolSupplier;
+import io.helidon.common.http.Http;
+import io.helidon.config.Config;
+import io.helidon.microprofile.cdi.RuntimeScoped;
+import io.helidon.webserver.Routing;
+import io.helidon.webserver.ServerConfiguration;
+import io.helidon.webserver.Service;
+import io.helidon.webserver.SocketConfiguration;
+import io.helidon.webserver.StaticContentSupport;
+import io.helidon.webserver.WebServer;
+import io.helidon.webserver.jersey.JerseySupport;
+
+import org.eclipse.microprofile.config.ConfigProvider;
 
 /**
- * Extension to gather JAX-RS application or JAX-RS resource classes
- * if no application is present.
+ * Extension to handle web server configuration and lifecycle.
  */
 public class ServerCdiExtension implements Extension {
-    private final List<Class<?>> resourceClasses = new LinkedList<>();
-    private final List<Class<? extends Application>> applications = new LinkedList<>();
+    static {
+        HelidonFeatures.register(HelidonFlavor.MP, "Server");
+    }
 
-    /**
-     * Gather Application or resource classes to start.
-     *
-     * @param pit injection target
-     * @param <T> any type
-     */
-    @SuppressWarnings("unchecked")
-    public <T> void gatherApplications(@Observes ProcessInjectionTarget<T> pit) {
-        AnnotatedType<T> at = pit.getAnnotatedType();
-        if (!at.isAnnotationPresent(ApplicationScoped.class)) {
+    private static final Logger LOGGER = Logger.getLogger(ServerCdiExtension.class.getName());
+
+    // build time
+    private ServerConfiguration.Builder serverConfigBuilder = ServerConfiguration.builder()
+            .port(7001);
+    private Routing.Builder routingBuilder = Routing.builder();
+    private Map<String, Routing.Builder> namedRoutings = new HashMap<>();
+
+    // configuration option that can be provided, only available in `startServer`
+    private Supplier<? extends ExecutorService> jaxRsExecutorService;
+    private String basePath;
+    private Config config;
+
+    // runtime
+    private WebServer webserver;
+    private int port;
+    private String listenHost = "0.0.0.0";
+
+    private void prepareRuntime(@Observes @Initialized(RuntimeScoped.class) Config config) {
```

Comment:
As observed elsewhere: what about a simple qualifier that doesn't commit you to scope annotations that aren't actually scope annotations?  Something like `@RuntimeInitializing`?

Reply:
The problem with that approach is that you need a combination of `Runtime` and `BuildTime` + initialization and finishing (and there is even pre-destroy and post-destroy).
Going with the "scope like" approach, I just need the two annotations to mark the "phase" and can use existing annotations to mark the lifecycle of the phase. If I use dedicated annotations, I need one for each combination (`RuntimeInitializing`, `BuildtimeFinish`).
I quite liked the coding with the current solution and I did not like the fully descriptive annotation (tried both).
I can change it again, though we may actually want to have beans that are in such a scope (e.g. creating real scope, at least for the `RuntimeScoped` scope), as that would make sense for some beans we use (such as data sources may be created in RuntimeScoped). 
I will leave this decision to you, as you understand the CDI paradigm the best from our team (or we can discuss on a meeting).

Reply:
We now have:
`@RuntimeInitializing` qualifier + `Config` as event
`@BuildTimeEvent.BuildTimeStart` + `BuildTimeEvent`
`@BuildTimeEvent.BuildTimeEnd` + `BuildTimeEvent`